scipy/io/matlab/mio.py

```diff
@@ -14,30 +14,38 @@
 __all__ = ['mat_reader_factory', 'loadmat', 'savemat', 'whosmat']
 
 
-def _open_file(file_like, appendmat):
+@contextmanager
+def _open_file_context(file_like, appendmat, mode='rb'):
+    f, opened = _open_file(file_like, appendmat, mode)
+    yield f
+    if opened:
+        f.close()
+
+
+def _open_file(file_like, appendmat, mode='rb'):
     """
     Open `file_like` and return as file-like object. First, check if object is
     already file-like; if so, return it as-is. Otherwise, try to pass it
     to open(). If that fails, and `file_like` is a string, and `appendmat` is true,
     append '.mat' and try again.
     """
-    try:
-        file_like.read(0)
+    if hasattr(file_like, 'read'):
```

Comment:
Sorry to be anal, but for completeness, how about:

```
reqs = {'read'} if set(mode) & set('r+') else {}
if set(mode) & set('wax+'):
    reqs.add('write')
if reqs.issubset(dir(file_like)):
    return file_like, False
```

Reply:
`reqs = {'read'} if set(mode) & set('r+') else {}` would create a dict if mode doesn't overlap with 'r+'.
But I've taken the suggestion on board.